documentation api (#48)
Cyberwatch/cyberwatch_api_toolbox: documentation.md

# Cyberwatch API Toolbox documentation

## Available methods

#### Ping

Send a GET request to `/api/v2/ping` to validate that the API is working.

###### Usage example and expected result:

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).ping()
INFO:root:OK
True
```

#### Servers

Send a GET request to `/api/v2/servers` to retrieve the list of all servers.

###### Usage example and expected result:

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).servers()
[<cbw_api_toolbox.cbw_objects.cbw_server.CBWServer ...]
```

#### Server

Send a GET request to `/api/v2/servers/{SERVER_ID}` to retrieve the information of a server.

##### Usage example and expected result:

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).server(SERVER_ID)
[<cbw_api_toolbox.cbw_objects.cbw_server.CBWServer]
```

#### Update server

Send a PATCH request  `/api/v2/servers/{SERVER_ID}` to update the information of a server.

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).update_server(SERVER_ID, INFO)
True
```

#### Delete server

> Send a DELETE request `/api/v2/servers/{SERVER_ID}` to delete SERVER_ID.

Reviewer: We delete a server not a `SERVER_ID` (same comment as the previous review)

=> You use the ID of a Server (called SERVER_ID) to find the server you want to delete....

> Send a DELETE request `/api/v2/servers/{SERVER_ID}` to delete the server of SERVER_ID.

Reviewer: let's make it simpler :
```suggestion
Send a DELETE request `/api/v2/servers/{SERVER_ID}` to delete a server.
```

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).delete_server(SERVER_ID)
True
```

#### Remote accesses

Send a GET request `/api/v2/remote_accesses` to retrieve the list of all remote accesses.

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).remote_accesses()
[<cbw_api_toolbox.cbw_objects.cbw_remote_access.CBWRemoteAccess...]
```

#### Remote access

Send a GET request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to retrieve the information of a particular remote access.

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).remote_access(REMOTE_ACCESS_ID)
[<cbw_api_toolbox.cbw_objects.cbw_remote_access.CBWRemoteAccess]
```

#### Create remote access

> Send a POST request `/api/v2/remote_accesses` to create a remote access

Reviewer: Sometimes you put a dot " . " as the end of the sentence, sometimes you don't
- put the dot on every sentence that requires one
- make sure this is applied to the entire file (i'm not commenting on every missing dot)

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).create_remote_access(INFO)
INFO:root:remote access successfully created {ADDRESS}
True
```

#### Update remote access

Send a PATCH request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to update the information of a particular remote access.

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).update_remote_access(REMOTE_ACCESS_ID, INFO)
True
```

#### Delete remote access

> Send a DELETE request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to delete remote access of REMOTE_ACCESS_ID.

Reviewer: ```suggestion
Send a DELETE request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to delete a remote access.
```

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).delete_remote_access(REMOTE_ACCESS_ID)
True
```

## Available objects and their attributes

### Server

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| id                        | String (hash) | Unique id of the computer         | 4a78524087574c12453dea248a91cadb                  |
| agent_version             | String        | Agent version used for supervision in case of agent supervision. Empty for agentless connections| DESKTOP-AAA111                                    |
| applications              | List          | List of the computer's applications|                                                  |
| boot_at                   | Date          | Date of the computer boot         | 2016-05-30T09:54:58.000+02:00                     |

> | category                  | String        | Computer category                 | server                                            |

Reviewer: ```suggestion
| category                  | String        | Computer's category                 | server                                            |
```

> | criticality               | String        | Computer criticality              | criticality_medium                                |

Reviewer: ```suggestion
| criticality               | String        | Computer's criticality              | criticality_medium                                |
```

> | cve_announcements         | List          | list of CVEs affecting the server | [CveAnnouncements](#Cve)                          |

Reviewer: ```suggestion
| cve_announcements         | List          | List of CVEs affecting the computer | [CVEs](#Cve)                          |
```

| cve_announcements_count   | Int           | Number of CVEs                    | 3                                                 |    

> | deploying_period          | Object        | Computer deploying period         | [DeployingPeriod](#Deploying-period)              |

Reviewer: ```suggestion
| deploying_period          | Object        | Computer's deploying period         | [Deploying Period](#Deploying-period)              |
```

> | description               | String        | Computer description              | "machine of production"                           |

Reviewer: ```suggestion
| description               | String        | Computer's description              | "production machine"                           |
```

> | groups                    | List          | Computer groups                   | [Group](#Group)                                   |

Reviewer: ```suggestion
| groups                    | List          | Computer's groups                   | [Group](#Group)                                   |
```

> | hostname                  | String        | Name of the machine               | "XXX.XXX.XXX.XXX"                                 |

Reviewer: ```suggestion
| hostname                  | String        | Computer's hostname               | "XXX.XXX.XXX.XXX"                                 |
```

> | ignoring_policy           | Object        | Ignoring policy of the computer   | [IgnoringPolicy](#Ignoring-policy)                |

Reviewer: ```suggestion
| ignoring_policy           | Object        | Computer's Ignoring policy | [Ignoring Policy](#Ignoring-policy)                |
```

> | last_communication        | String        | Last communication of the computer| "2019-04-08T15:46:22.000+02:00"                   |

Reviewer: ```suggestion
| last_communication        | String        | Date of last communication with the computer | "2019-04-08T15:46:22.000+02:00"                   |
```

> | os                        | Object        | Os of the computer                | [Os](#Os)                                         |

Reviewer: ```suggestion
| os                        | Object        | Computer's OS                | [Os](#Os)                                         |
```

| packages                  | Object        | List of the computer's packages   | [Packages](#Packages)                             |
| reboot_required           | Bool          | The computer requires to be rebooted| false            |

> | remote_ip                 | String        | Computer IP address               | XXX.XXX.XXX.XXX                                   |

Reviewer: ```suggestion
| remote_ip                 | String        | Computer's IP address               | XXX.XXX.XXX.XXX                                   |
```

| security_announcements    | Object        | List of the computer's security announcements| [SecurityAnnouncements](#Security-announcements)|

> | status                    | List          | Computer status                   | comment: "Communication failure"                  |

Reviewer: ```suggestion
| status                    | List          | Computer's status                   | comment: "Communication failure"                  |
```

| updates                   | Object        | Available updates for the computer| [Updates](#Updates)                               |
| updates_count             | Int           | Number of available updates for the computer| 3                                       |

### Group

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| id                        | Int           | id of the group                   | 140                                               |
| name                      | String        | name of the group                 | "app_windows"                                     |

> | created_at                | String (date) | Date of created                   | "2018-07-23T09:41:31.000+02:00"                   |

Reviewer: ```suggestion
| created_at                | String (date) | Date of creation                   | "2018-07-23T09:41:31.000+02:00"                   |
```

> | updated_at                | String (date) | Date of updated                   | "2018-07-23T09:41:31.000+02:00"                   |

Reviewer: ```suggestion
| updated_at                | String (date) | Date of last update                   | "2018-07-23T09:41:31.000+02:00"                   |
```

### Ignoring policy

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| ignoring_policy_items     | Object        | Items of the ignoring policy      | [IgnoringPolicyItems](#Ignoring-policy-items)     |
| name                      | String        | Name of the ignoring policy       | "test_policy"                                     |

### Ignoring policy items

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| keyword                   | String        | Keyword of the packages           | ""                                                |
| version                   | String        | Version of the packages           | ""                                                |

> ### Packages

Reviewer: ```suggestion
### Package
```

the object class name has no 's'

| created_at                | String (date) | Creation date of the CVE in Cyberwatch| "2019-04-08T22:17:34.000+02:00"               |
| updated_at                | String (date) | Last modification date of the CVE in Cyberwatch| "2019-04-29T09:33:37.000+02:00"      |

#### Current

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| vendor                    | String        | Package vendor                    | null                                              |
| product                   | String        | Package name                      | "libsystemd0"                                     |
| hash_index                | String (hash) | Hash identifier                   | "13a61dd2bfe0fbb063a97b66c6234759"                |
| type                      | String        | Package type in Cyberwatch        | "Packages::Deb"                                   |

> | version                   | String        | Package version                   | "237-3ubuntu10.13"                                |

Reviewer: Missing documentation of the `Node` object : https://github.com/Cyberwatch/cyberwatch_api_toolbox/blob/e532605342d0c1b0a19f1ede63559952c3a25ad5/cbw_api_toolbox/cbw_objects/cbw_node.py

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).delete_server(SERVER_ID)
>>>True
```

#### Remote accesses

Send a GET request `/api/v2/remote_accesses` to retrieve the list of all remote accesses.

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).remote_accesses()
[<cbw_api_toolbox.cbw_objects.cbw_remote_access.CBWRemoteAccess...]
```

#### Remote access

Send a GET request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to retrieve the information of REMOTE_ACCESS_ID.

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).remote_access(REMOTE_ACCESS_ID)
[<cbw_api_toolbox.cbw_objects.cbw_remote_access.CBWRemoteAccess]
```

#### Create remote access

Send a POST request `/api/v2/remote_accesses` to create a remote access

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).create_remote_access(INFO)
INFO:root:remote access successfully created {ADDRESS}
True
```

#### Update remote access

Send a PATCH request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to update the information of REMOTE_ACCESS_ID

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).update_remote_access(REMOTE_ACCESS_ID, INFO)
True
```

#### Delete remote access

> Send a DELETE request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to delete REMOTE_ACCESS_ID

Reviewer: not resolved...

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).delete_remote_access(REMOTE_ACCESS_ID)
True
```

## Available objects and their attributes

> #### Server

Reviewer: For objects and their attributes, I believe :
- We should write the name of the object as it is in the code (eg. `Server` => `CBWServer`) as it is clearer and avoids confusion with methods
- The attributes listed in the documentation should match exactly the object's attributes (doesn't seem to be the case right now, at least with the CBWCve object here... see comments below about the object Server)

I didn't review and comment on every objects, but these changes should be applied to all documented objects

> 

Reviewer: not required \n

> | Attribute                 | Type          | Description                       | Exemple of possible value                         |

Reviewer: not resolved...

| os                        | Object        | Os of the machine                 | [Os](#Os)                                         |
| packages                  | Object        | List of the computer's packages   | [Packages](#Packages)                             |
| reboot_required           | Bool          | obligation to restart the machine | false                                             |
| remote_ip                 | String        | IP of the machine                 | XXX.XXX.XXX.XXX                                   |
| security_announcements    | Object        | List of the computer's security announcements| [SecurityAnnouncements](#Security-announcements)|
| status                    | List          | Computer status                   | comment: "Communication failure"                  |
| updates                   | Object        | Available updates for the computer| [Updates](#Updates)                               |
| updates_count             | Int           | Number of available updates for the computer| 3                                       |

> 

Reviewer: not required \n

#### Deploying period

> | Attribute                 | Type          | Description                       | Exemple of possible value                         |

Reviewer: same, not resolved..

|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| name                      | String        | Name of deploying period          | "test"                                            |
| autoplanning              | Bool          | Planning of the deployment        | false                                             |
| autoreboot                | Bool          | reboot update was performed       | false                                             |

> | start_time                | String        | Time start of deployment          | "00:00"                                           |

Reviewer: please resolve this comment

> | end_time                  | String        | End of deployment                 | "06:00"                                           |

Reviewer: please resolve this comment

> | next_occurence            | String (date) | Date of the next deployment       | "2019-04-27T00:00:00.000+02:00"                   |

Reviewer: please resolve this comment

#### Os

| Attribute                 | Type          | Description                       | Exemple of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|

> | key                       | String        | Key of the os                     | "ubuntu_1804_64"                                  |

Reviewer: please resolve this comment

> | name                      | String        | Name of Os                        | "Ubuntu 18.04"                                    |

Reviewer: please resolve this comment

| arch                      | String        | Arhitecture of Os                 | "x86_64"                                          |
| eol                       | String        | End of support of the os          | "2023-04-26T02:00:00.000+02:00"                   |
| short_name                | String        | Short name of the os              | "Ubuntu 18.04"                                    |
| type                      | String        | Os type                           | "Os::Ubuntu"                                      |
| hash_index                | String (hash) | Hash identifier                   | "c3f4de66dd34cec0156bb439002bc4..."               |
| created_at                | String (date) | Date of created                   | "2019-04-08T11:01:20.000+02:00"                   |
| updated_at                | String (date) | Date of updated                   | "2019-04-08T11:01:20.000+02:00"                   |

#### Ignoring policy

> | Attribute                 | Type          | Description                       | Exemple of possible value                         |

Reviewer 2: ```suggestion
| Attribute                 | Type          | Description                       | Example of possible value                         |
```

|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| ignoring_policy_items     | Object        | Items of the ignoring policy      | [IgnoringPolicyItems](#Ignoring-policy-items)     |

> | name                      | String        | Name of ignoring policy           | "test_policy"                                     |

Reviewer 2: ```suggestion
| name                      | String        | Name of the ignoring policy           | "test_policy"                                     |
```

#### Ignoring policy items

> | Attribute                 | Type          | Description                       | Exemple of possible value                         |

Reviewer 2: ```suggestion
| Attribute                 | Type          | Description                       | Example of possible value                         |
```

|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| keyword                   | String        | Keyword of the packages           | ""                                                |
| version                   | String        | Version of the packages           | ""                                                |

#### Packages

> | Attribute                 | Type          | Description                       | Exemple of possible value                         |

Reviewer 2: ```suggestion
| Attribute                 | Type          | Description                       | Example of possible value                         |
```

|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|

> | vendor                    | NoneType      | Vendor of the packages            | null                                              |

Reviewer 2: ```suggestion
| vendor                    | NoneType      | Vendor of the package            | null                                              |
```

Reviewer 2: Why NoneType ?

> | product                   | String        | Name of packages                  | "gpg-wks-server"                                  |

Reviewer 2: ```suggestion
| product                   | String        | Name of the package                  | "gpg-wks-server"                                  |
```

| hash_index                | String (hash) | Hash identifier                   | "0160e4a3d38518a2660ff61b89c1407df"               |

> | type                      | String        | Packages type                     | "Packages::Deb"                                   |

Reviewer 2: ```suggestion
| type                      | String        | Package type in Cyberwatch                     | "Packages::Deb"                                   |
```

> | version                   | String        | Version of packages               | "2.2.4-1ubuntu1.2"                                |

Reviewer 2: ```suggestion
| version                   | String        | Package version               | "2.2.4-1ubuntu1.2"                                |
```

#### Security announcements

> | Attribute                 | Type          | Description                       | Exemple of possible value                         |

Reviewer 2: ```suggestion
| Attribute                 | Type          | Description                       | Example of possible value                         |
```

|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|

> | type                      | String        | Security type                     | "SecurityAnnouncements::Ubuntu"                   |

Reviewer 2: ```suggestion
| type                      | String        | Security announcement type in Cyberwatch                     | "SecurityAnnouncements::Ubuntu"                   |
```

> | sa_code                   | String        | Security alert code               | "USN-3938-1"                                      |

Reviewer 2: ```suggestion
| sa_code                   | String        | Security announcement code               | "USN-3938-1"                                      |
```

> | created_at                | String (date) | Date of created                   | "2019-04-08T02:00:00.000+02:00"                   |

Reviewer 2: ```suggestion
| created_at                | String (date) | Creation date in Cyberwatch                   | "2019-04-08T02:00:00.000+02:00"                   |
```

> | updated_at                | String (date) | Date of updated                   | "2019-04-10T09:20:07.000+02:00"                   |

Reviewer 2: ```suggestion
| updated_at                | String (date) | Last modification date in Cyberwatch                   | "2019-04-10T09:20:07.000+02:00"                   |
```

> | link                      | String        | Link of the security              | "https://usn.ubuntu.com/3938-1"                   |

Reviewer 2: ```suggestion
| link                      | String        | URL of the security announcement              | "https://usn.ubuntu.com/3938-1"                   |
```

> | cve_code                  | String        | CVE Code public                   | "CVE-2019-3842"                                   |

Reviewer 2: ```suggestion
| cve_code                  | String        | CVE code                 | "CVE-2019-3842"                                   |
```

#### Updates

> | Attribute                 | Type          | Description                       | Exemple of possible value                         |

Reviewer 2: ```suggestion
| Attribute                 | Type          | Description                       | Example of possible value                         |
```

|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|

> | ignored                   | Bool          | Ignore of update                  | false                                             |

Reviewer 2: ```suggestion
| ignored                   | Bool          | boolean checking whether if the update has been ignored or not                  | false                                             |
```

> | patchable                 | Bool          | Update is possible                | true                                              |

Reviewer 2: ```suggestion
| patchable                 | Bool          | boolean checking whether if the update is patchable via Cyberwatch or not                | true                                              |
```

> | cve_announcements         | Object        | All the update availbale cve      | [CveAnnouncements](#Cve-announcements)            |

Reviewer 2: ```suggestion
| cve_announcements         | Object        | List of the CVEs concerned by the update      | [CveAnnouncements](#Cve-announcements)            |
```

> | current                   | Object        | Current version of packages       | [Current](#Current)                               |

Reviewer 2: ```suggestion
| current                   | Object        | Vulnerable version of the package affected by the update       | [Current](#Current)                               |
```

> | target                    | Object        | Target version of packages        | [Target](#Target)                                 |

Reviewer 2: ```suggestion
| target                    | Object        | Version of the package to update to      | [Target](#Target)                                 |
```

> | created_at                | String (date) | Date of created                   | "2019-04-09T17:13:07.000+02:00"                   |

Reviewer 2: ```suggestion
| created_at                | String (date) | Creation date of the update in Cyberwatch                   | "2019-04-09T17:13:07.000+02:00"                   |
```

> | updated_at                | String (date) | Date of updated                   | "2019-04-09T17:13:07.000+02:00"                   |

Reviewer 2: ```suggestion
| updated_at                | String (date) | Last modification date of the update in Cyberwatch                   | "2019-04-09T17:13:07.000+02:00"                   |
```

> #### Cve announcements

Reviewer 2: L'endpoint pour cet objet n'est pas indiqué

| Attribute                 | Type          | Description                       | Exemple of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|

> | id_cve                    | Int           | Id Cyberwatch of the vulnerability| 22                                                |

Reviewer 2: C'est pas plutôt id tout court ?

Reviewer 2: ```suggestion
| id                    | Int           | Unique id of the CVE in Cyberwatch | 22                                                |
```

> | id_cve                    | Int           | Id Cyberwatch of the vulnerability| 22                                                |

Reviewer: same no attribute id_cve (see below for server_id)

> | server_id                 | Int           | Id Cyberwatch of the server       | 21                                                |

Reviewer 2: Pourquoi server_id dans CveAnnouncement ?

> | server_id                 | Int           | Id Cyberwatch of the server       | 21                                                |

Reviewer: does the `cve_announcements` object even has an attribute `server_id` ?

I'm getting : 
```python
>>> print(serv.cve_announcements[0].server_id)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'CBWCve' object has no attribute 'server_id'
```

and the code of the class cbw_cve doesn't seem to have one either :
```python
"""CVE Model"""

class CBWCve:
    """CVE Model"""

    def __init__(self,
                 content="",
                 created_at="",
                 cve_code="",
                 cve_score="",
                 last_modified="",
                 published="",
                 updated_at="",
                 **kwargs): # pylint: disable=unused-argument
        self.content = content
        self.created_at = created_at
        self.cve_code = cve_code
        self.cve_score = cve_score
        self.last_modified = last_modified
        self.published = published
        self.updated_at = updated_at
```

> | target_id                 | Int           | Id target cve                     | 239309                                            |

Reviewer 2: Pourquoi target_id dans CveAnnouncement ?

> | target_id                 | Int           | Id target cve                     | 239309                                            |

Reviewer: same... or am I missing something ?

> | current_id                | Int           | Id current cve                    | 174477                                            |

Reviewer 2: Pourquoi current_id dans CveAnnouncement ?

> | created_at                | String (date) | Date of created                   | "2019-04-09T17:13:07.000+02:00"                   |

Reviewer 2: ```suggestion
| created_at                | String (date) | Creation date in Cyberwatch                   | "2019-04-09T17:13:07.000+02:00"                   |
```

> | updated_at                | String (date) | Date of updated                   | "2019-04-09T17:13:07.000+02:00"                   |

Reviewer 2: ```suggestion
| updated_at                | String (date) | Last modification date in Cyberwatch                   | "2019-04-09T17:13:07.000+02:00"                   |
```

> | ignored                   | Bool          | Ignore of CVE                     | false                                             |

Reviewer 2: ```suggestion
| ignored                   | Bool          | Boolean checking whether the CVE has been ignored on Cyberwatch or not                     | false                                             |
```

> | patchable                 | Bool          | Update is possible                | true                                              |

Reviewer 2: ```suggestion
| patchable                 | Bool          | Boolean checking whether the CVE is patchable on Cyberwatch or not                | true                                              |
```

> | cve_code                  | String        | Cve code for packages             | "CVE-2019-5953"                                   |

Reviewer 2: ```suggestion
| cve_code                  | String        | CVE code             | "CVE-2019-5953"                                   |
```

> | content                   | String        | Description of the cve code       | "In systemd before v242-r…her than \"allow_any\"."|

Reviewer 2: ```suggestion
| content                   | String        | CVE description       | "In systemd before v242-r…her than \"allow_any\"."|
```

> | published                 | String        | Date of publication               | "2019-04-09T23:29:03.000+02:00"                   |

Reviewer 2: ```suggestion
| published                 | String        | Publication date of the CVE by the authorities               | "2019-04-09T23:29:03.000+02:00"                   |
```

> | last_modified             | String        | Last modification                 | "2019-04-26T16:45:22.000+02:00"                   |

Reviewer 2: ```suggestion
| last_modified             | String        | Last modification date of the CVE by the authorities                 | "2019-04-26T16:45:22.000+02:00"                   |
```

> | cve_score                 | Float         | score of vulnerablity according to cve| 4.4                                           |

Reviewer 2: ```suggestion
| cve_score                 | Float         | CVE score | 4.4                                           |
```

> | created_at                | String (date) | Date of created                   | "2019-04-08T22:17:34.000+02:00"                   |

Reviewer 2: ```suggestion
| created_at                | String (date) | Creation date of the CVE in Cyberwatch                   | "2019-04-08T22:17:34.000+02:00"                   |
```

> | updated_at                | String (date) | Date of updated                   | "2019-04-29T09:33:37.000+02:00"                   |

Reviewer 2: ```suggestion
| updated_at                | String (date) | Last modification date of the CVE in Cyberwatch                   | "2019-04-29T09:33:37.000+02:00"                   |
```

#### Current

> | Attribute                 | Type          | Description                       | Exemple of possible value                         |

Reviewer 2: ```suggestion
| Attribute                 | Type          | Description                       | Example of possible value                         |
```

|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|

> | vendor                    | NoneType      | Vendor of the packages            | null                                              |

Reviewer 2: Why NoneType ?

Reviewer 2: ```suggestion
| vendor                    | NoneType      | Package vendor            | null                                              |
```

> | product                   | String        | Product of packages               | "libsystemd0"                                     |

Reviewer 2: ```suggestion
| product                   | String        | Package name               | "libsystemd0"                                     |
```

| hash_index                | String (hash) | Hash identifier                   | "13a61dd2bfe0fbb063a97b66c6234759"                |

> | type                      | String        | Current packages type             | "Packages::Deb"                                   |

Reviewer 2: ```suggestion
| type                      | String        | Package type in Cyberwatch             | "Packages::Deb"                                   |
```

> | version                   | String        | Version of product                | "237-3ubuntu10.13"                                |

Reviewer 2: ```suggestion
| version                   | String        | Package version                | "237-3ubuntu10.13"                                |
```

#### Target

> 

Reviewer: not required \n

> | Attribute                 | Type          | Description                       | Exemple of possible value                         |

Reviewer 2: Ce tableau est exactement le même que `Current`. Le supprimer et changer le titre au-dessus en `Current / Target`

# Cyberwatch API Toolbox documentation

## Available methods

#### Ping

Send a GET request to `/api/v2/ping` to validate that the API is working.

###### Usage example and expected result:

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).ping()
INFO:root:OK
True
```

#### Servers

Send a GET request to `/api/v2/servers` to retrieve the list of all servers.

###### Usage example and expected result:

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).servers()
[<cbw_api_toolbox.cbw_objects.cbw_server.CBWServer ...]
```

#### Server

Send a GET request to `/api/v2/servers/{SERVER_ID}` to retrieve the information of a server.

> ##### Usage example and expected result:

Reviewer 3: Ajouter #

```python
>>> CBWApi(URL, API_KEY, SECRET_KEY).server(SERVER_ID)
[<cbw_api_toolbox.cbw_objects.cbw_server.CBWServer]
```

#### Update server

Send a PATCH request  `/api/v2/servers/{SERVER_ID}` to update the information of a server.

> ```python

Reviewer 3: ###### Usage example and expected result:

Reviewer 3: C'est pas résolu, c'est pas ce que j'ai demandé, et c'est pas vérifié car c'est syntaxiquement non valide, et en plus c'est moche. :rage:

>>> CBWApi(URL, API_KEY, SECRET_KEY).update_server(SERVER_ID, INFO)
True
```

#### Delete server

Send a DELETE request `/api/v2/servers/{SERVER_ID}` to delete a server.

> ```python

Reviewer 3: ###### Usage example and expected result:

Reviewer 3: no resolu

>>> CBWApi(URL, API_KEY, SECRET_KEY).delete_server(SERVER_ID)
True
```

#### Remote accesses

Send a GET request `/api/v2/remote_accesses` to retrieve the list of all remote accesses.

> ```python

Reviewer 3: ###### Usage example and expected result:

>>> CBWApi(URL, API_KEY, SECRET_KEY).remote_accesses()
[<cbw_api_toolbox.cbw_objects.cbw_remote_access.CBWRemoteAccess...]
```

#### Remote access

Send a GET request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to retrieve the information of a particular remote access.

> ```python

Reviewer 3: ###### Usage example and expected result:

>>> CBWApi(URL, API_KEY, SECRET_KEY).remote_access(REMOTE_ACCESS_ID)
[<cbw_api_toolbox.cbw_objects.cbw_remote_access.CBWRemoteAccess]
```

#### Create remote access

Send a POST request `/api/v2/remote_accesses` to create a remote access.

> ```python

Reviewer 3: ###### Usage example and expected result:

Reviewer 3: no resolu

>>> CBWApi(URL, API_KEY, SECRET_KEY).create_remote_access(INFO)
INFO:root:remote access successfully created {ADDRESS}
True
```

#### Update remote access

Send a PATCH request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to update the information of a particular remote access.

> ```python

Reviewer 3: ###### Usage example and expected result:

Reviewer 3: no resolu

>>> CBWApi(URL, API_KEY, SECRET_KEY).update_remote_access(REMOTE_ACCESS_ID, INFO)
True
```

#### Delete remote access

Send a DELETE request `/api/v2/remote_accesses/{REMOTE_ACCESS_ID}` to delete a remote access.

> ```python

Reviewer 3: ###### Usage example and expected result:

Reviewer 3: no resolu

| cve_announcements_count   | Int           | Number of CVEs                    | 3                                                 |    
| deploying_period          | Object        | Computer's deploying period       | [Deploying Period](#Deploying-period)             |
| description               | String        | Computer description              | "machine of production"                           |
| groups                    | List          | Computer groups                   | [Group](#Group)                                   |
| hostname                  | String        | Name of the machine               | "XXX.XXX.XXX.XXX"                                 |
| ignoring_policy           | Object        | Computer's Ignoring policy        | [Ignoring Policy](#Ignoring-policy)               |
| last_communication        | String        | Date of last communication with the computer| "2019-04-08T15:46:22.000+02:00"         |
| os                        | Object        | Os of the computer                | [Os](#Os)                                         |

> | packages                  | Object        | List of the computer's packages   | [Packages](#Packages)                             |

Reviewer: ```suggestion
| packages                  | Object        | List of the computer's packages   | [Packages](#Package)                             |
```

#### Os

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| key                       | String        | OS key in Cyberwatch              | "ubuntu_1804_64"                                  |
| name                      | String        | OS name                           | "Ubuntu 18.04"                                    |
| arch                      | String        | OS architecture                   | "x86_64"                                          |
| eol                       | String        | OS end of support as provided by their vendor          | "2023-04-26T02:00:00.000+02:00"                   |
| short_name                | String        | OS short name                     | "Ubuntu 18.04"                                    |
| type                      | String        | OS type in Cyberwatch             | "Os::Ubuntu"                                      |

> | hash_index                | String (hash) | Hash identifier                   | "c3f4de66dd34cec0156bb439002bc4..."               |

Reviewer 3: Le champ hash_index n'existe pas

| created_at                | String (date) | Creation date in Cyberwatch       | "2019-04-08T11:01:20.000+02:00"                   |
| updated_at                | String (date) | Last modification date in Cyberwatch| "2019-04-08T11:01:20.000+02:00"                   |

### Ignoring policy

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| ignoring_policy_items     | Object        | Items of the ignoring policy      | [Ignoring Policy Items](#Ignoring-policy-items)     |
| name                      | String        | Name of the ignoring policy       | "test_policy"                                     |

### Ignoring policy items

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|

> | keyword                   | String        | Keyword of the packages           | ""                                                |

Reviewer: Is it intentional to leave the field example empty ?

> | version                   | String        | Version of the packages           | ""                                                |

Reviewer: Is it intentional to leave the field example empty ?

### Package

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|

> | vendor                    | String        | Vendor of the package             | null                                              |

Reviewer: The attribute vendor is always null ? Because the example makes it feel like it

| product                   | String        | Name of the package               | "gpg-wks-server"                                  |
| hash_index                | String (hash) | Hash identifier                   | "0160e4a3d38518a2660ff61b89c1407df"               |
| type                      | String        | Package type in Cyberwatch        | "Packages::Deb"                                   |
| version                   | String        | Package version                   | "2.2.4-1ubuntu1.2"                                |

#### Security announcements

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|

> | type                      | String        | Security announcement type in Cyberwatch| "SecurityAnnouncements::Ubuntu"             |

Reviewer 3: Pour tous les attributs respecter l'ordre alphabétique

| sa_code                   | String        | Security announcement code        | "USN-3938-1"                                      |
| created_at                | String (date) | Creation date in Cyberwatch       | "2019-04-08T02:00:00.000+02:00"                   |
| updated_at                | String (date) | Last modification date in Cyberwatch| "2019-04-10T09:20:07.000+02:00"                 |
| link                      | String        | URL of the security announcement  | "https://usn.ubuntu.com/3938-1"                   |
| cve_code                  | String        | CVE Code                          | "CVE-2019-3842"                                   |

#### Updates

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| ignored                   | Bool          | Boolean checking whether if the update has been ignored or not| false                 |

> | patchable                 | Bool          | Boolean checking whether if the update is patchable via Cyberwatch or not| true       |

Reviewer 3: Pour tous les attributs respecter l'ordre alphabétique

| current                   | Object        | Vulnerable version of the package affected by the update| [Current](#Current)         |
| target                    | Object        | Version of the package to update to| [Target](#Current)                               |
| created_at                | String (date) | Creation date of the update in Cyberwatch| "2019-04-09T17:13:07.000+02:00"            |
| updated_at                | String (date) | Last modification date of the update in Cyberwatch| "2019-04-09T17:13:07.000+02:00"   |

### Cve

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| cve_code                  | String        | CVE code                          | "CVE-2019-5953"                                   |
| content                   | String        | CVE description                   | "In systemd before v242-r…her than \"allow_any\"."|

> | published                 | String        | Publication date of the CVE by the authorities| "2019-04-09T23:29:03.000+02:00"       |

Reviewer 3: Pour tous les attributs respecter l'ordre alphabétique

| last_modified             | String        | Last modification date of the CVE by the authorities| "2019-04-26T16:45:22.000+02:00" |
| cve_score                 | Float         | CVE score                         | 4.4                                               |
| created_at                | String (date) | Creation date of the CVE in Cyberwatch| "2019-04-08T22:17:34.000+02:00"               |
| updated_at                | String (date) | Last modification date of the CVE in Cyberwatch| "2019-04-29T09:33:37.000+02:00"      |

#### Current

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|

> | vendor                    | String        | Package vendor                    | null                                              |

Reviewer 3: Pour tous les attributs respecter l'ordre alphabétique

| type                      | String        | Package type in Cyberwatch        | "Packages::Deb"                                   |
| version                   | String        | Package version                   | "237-3ubuntu10.13"                                |

### Node

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| id                        | String        | Node Id                           | 1                                                 |
| name                      | String        | Node name                         | "master"                                          |
| description               | String        | Node description                  | "Node master"                                     |
| url                       | String        | Node Url                          | "https://cyberwatch.fr"                           |

> | olympe_version            | String        | Olympe version of node            | "9.6.2"                                           |

Reviewer: ```suggestion
| olympe_version            | String        | Version of Cyberwatch application installed on the node            | "9.6.2"                                           |
```

| cbw_on_premise_version    | String        | Cbw_on_premsise version           | "2.6"                                             |
| boot_time                 | String        | Node boot time                    | "06 mai 12h02"                                    |

> | created_at                | String        | Creation date of node             | "2019-04-08T02:00:00.000+02:00"                   |

Reviewer: ```suggestion
| created_at                | String        | Creation date of the node             | "2019-04-08T02:00:00.000+02:00"                   |
```

> | updated_at                | String        | Last modification of node         | "2019-04-08T02:00:00.000+02:00"                   |

Reviewer: ```suggestion
| updated_at                | String        | Last modification of the node         | "2019-04-08T02:00:00.000+02:00"                   |
```

| name                      | String        | OS name                           | "Ubuntu 18.04"                                    |
| short_name                | String        | OS short name                     | "Ubuntu 18.04"                                    |
| type                      | String        | OS type in Cyberwatch             | "Os::Ubuntu"                                      |
| updated_at                | String (date) | Last modification date in Cyberwatch| "2019-04-08T11:01:20.000+02:00"                 |

### Ignoring policy

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| ignoring_policy_items     | Object        | Items of the ignoring policy      | [Ignoring Policy Items](#Ignoring-policy-items)   |
| name                      | String        | Name of the ignoring policy       | "test_policy"                                     |

> ### Ignoring policy items

Reviewer 3: item sans s

| version                   | String        | Version of the packages           | "0.1"                                             |

### Package

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| hash_index                | String (hash) | Hash identifier                   | "0160e4a3d38518a2660ff61b89c1407df"               |
| product                   | String        | Name of the package               | "gpg-wks-server"                                  |
| type                      | String        | Package type in Cyberwatch        | "Packages::Deb"                                   |
| vendor                    | String        | Vendor of the package             | null                                              |
| version                   | String        | Package version                   | "2.2.4-1ubuntu1.2"                                |

> #### Security announcements

Reviewer 3: Titre non uniforme 

![image](https://user-images.githubusercontent.com/4974759/57314598-4398e180-70f2-11e9-89d2-f4888cdf9da4.png)

Reviewer 3: En plus il y a pas de S a announcements pour rester cohérent

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| created_at                | String (date) | Creation date in Cyberwatch       | "2019-04-08T02:00:00.000+02:00"                   |
| cve_code                  | String        | CVE Code                          | "CVE-2019-3842"                                   |
| link                      | String        | URL of the security announcement  | "https://usn.ubuntu.com/3938-1"                   |
| sa_code                   | String        | Security announcement code        | "USN-3938-1"                                      |
| type                      | String        | Security announcement type in Cyberwatch| "SecurityAnnouncements::Ubuntu"             |
| updated_at                | String (date) | Last modification date in Cyberwatch| "2019-04-10T09:20:07.000+02:00"                 |

> #### Updates

Reviewer 3: Update sans s

| Attribute                 | Type          | Description                       | Example of possible value                         |
|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| created_at                | String (date) | Creation date of the update in Cyberwatch| "2019-04-09T17:13:07.000+02:00"            |
| current                   | Object        | Vulnerable version of the package affected by the update| [Current](#Current)         |

> | cve_announcements         | Object        | List of the CVEs concerned by the update| [Cve Announcements](#Cve-announcements)     |

Reviewer 3: `[Cve Announcements](#Cve-announcements)` lien cassé :-1:

| ignored                   | Bool          | Boolean checking whether if the update has been ignored or not| false                 |
| patchable                 | Bool          | Boolean checking whether if the update is patchable via Cyberwatch or not| true       |

> | target                    | Object        | Version of the package to update to| [Target](#Current)                               |

Reviewer 3: [Target](#Current)   :1st_place_medal:

| updated_at                | String (date) | Last modification date of the update in Cyberwatch| "2019-04-09T17:13:07.000+02:00"   |

> ### Cve

Reviewer 3: Cve Announcement

|---------------------------|:-------------:|:---------------------------------:|---------------------------------------------------|
| content                   | String        | CVE description                   | "In systemd before v242-r…her than \"allow_any\"."|
| created_at                | String (date) | Creation date of the CVE in Cyberwatch| "2019-04-08T22:17:34.000+02:00"               |
| cve_code                  | String        | CVE code                          | "CVE-2019-5953"                                   |
| cve_score                 | Float         | CVE score                         | 4.4                                               |
| last_modified             | String        | Last modification date of the CVE by the authorities| "2019-04-26T16:45:22.000+02:00" |
| published                 | String        | Publication date of the CVE by the authorities| "2019-04-09T23:29:03.000+02:00"       |
| updated_at                | String (date) | Last modification date of the CVE in Cyberwatch| "2019-04-29T09:33:37.000+02:00"      |

> #### Current

Reviewer 3: Devrait être Package
